typescript/infra/config/registry.ts

```diff
@@ -0,0 +1,102 @@
+import { dirname, join } from 'path';
+import { fileURLToPath } from 'url';
+
+import { ChainAddresses } from '@hyperlane-xyz/registry';
+import { LocalRegistry } from '@hyperlane-xyz/registry/local';
+import {
+  ChainMap,
+  ChainMetadata,
+  ChainName,
+  getDomainId as resolveDomainId,
+  getReorgPeriod as resolveReorgPeriod,
+} from '@hyperlane-xyz/sdk';
+import { objFilter, rootLogger } from '@hyperlane-xyz/utils';
+
+import { DeployEnvironment } from '../src/config/environment.js';
+
+const DEFAULT_REGISTRY_URI = join(
+  dirname(fileURLToPath(import.meta.url)),
+  '../../../../',
+  'hyperlane-registry',
+);
+
+// A global Registry singleton
+// All uses of chain metadata or chain address artifacts should go through this registry.
+let registry: LocalRegistry;
+
+export function setRegistry(reg: LocalRegistry) {
+  registry = reg;
+}
+
+export function getRegistry(): LocalRegistry {
+  if (!registry) {
+    const registryUri = process.env.REGISTRY_URI || DEFAULT_REGISTRY_URI;
+    rootLogger.info('Using registry URI:', registryUri);
+    registry = new LocalRegistry({
+      uri: registryUri,
+      logger: rootLogger.child({ module: 'infra-registry' }),
+    });
+  }
+  return registry;
+}
+
+export function getChains(): ChainName[] {
+  return getRegistry().getChains();
+}
+
+export function getChain(chainName: ChainName): ChainMetadata {
+  return getRegistry().getChainMetadata(chainName);
+}
+
+export function getDomainId(chainName: ChainName): number {
+  return resolveDomainId(getChain(chainName));
+}
+
+export function getReorgPeriod(chainName: ChainName): number {
+  return resolveReorgPeriod(getChain(chainName));
+}
+
+export function getChainMetadata(): ChainMap<ChainMetadata> {
+  return getRegistry().getMetadata();
+}
+
+export function getChainAddresses(): ChainMap<ChainAddresses> {
+  return getRegistry().getAddresses();
+}
+
+export function getEnvChains(env: DeployEnvironment): ChainName[] {
```

Comment:
I'm a little worried that this could lead to confusion / weird things with this function possibly returning a different set of chains than the list that's in the environment-specific configs (e.g. the `supportedChains` here in mainnet3 https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3646/files#diff-966e3e77cfd1880c5d9e2e5b69c9b9d8526b8d9006f3766ec21d33f226afba5dR7)

Is this intended to be used in a different way, e.g. instead of getting the list of chains that we want mainnet3 infrastructure to be aware of (e.g. "core" chains that we may wanna run validators / relayers for), this is just to get all "mainnet" chains in the registry? Wondering if there's any way to make the different more clear